mbuild/lattice.py

```diff
@@ -0,0 +1,401 @@
+              isinstance(basis_vectors, tuple)):
+            for lst in basis_vectors:
+                if len(lst) != 2:
+                    raise ValueError('Too many arguments per basis vector. '
+                                     '{} was provided, but the format should '
+                                     'be: vector = ( (\'ID1\' , [x1,y1,z1]), '
+                                     '( \'ID2\' , [x2,y2,z2]) )'.format(lst))
+                elif not isinstance(lst[0], str):
+                    raise TypeError('ID Error, not a string. {} is not of '
+                                    'type str. Please correct in the '
+                                    'basis_vectors input {}.'
+                                    .format(lst[0], basis_vectors))
+
+                if len(lst[1]) == dimension:
+                    if isinstance(lst[1], tuple) or isinstance(lst[1], list):
+                        for i in range(dimension):
+                            if (lst[1][i] < 0 or lst[1][i] >= 1):
+                                raise ValueError('Incorrect basis fractions. '
+                                                 '{} is supposed be 0 >= and '
+                                                 '< 1.'.format(lst[1]))
+
+                    else:
+                        raise TypeError('Incorrect Format: expected <list> '
+                                        'or <tuple>. {} was passed in as '
+                                        ' type: {} from basis_vectors.'
+                                        .format(lst[1], type(lst[1])))
+                else:
+                    raise ValueError('Basis Vector does not match dimension. '
+                                     '{} does not have vector components '
+                                     'equal to the dimension specified: {}'
+                                     .format(lst[1], dimension))
+        else:
+            raise TypeError('Incorrect Type: "basis_vectors" is of type {}, '
+                            'not type <list> or <tuple>. Please review '
+                            'the documentation for proper format.'
+                            .format(type(basis_vectors)))
+
+        if not isinstance(basis_vectors, defaultdict):
+            overlap_dict = defaultdict(list)
+            for the_id, vector in basis_vectors:
+                num_iterations = 3
+                if dimension == 3:
+                    for x in range(num_iterations):
+                        for y in range(num_iterations):
+                            for z in range(num_iterations):
+                                tmpx = vector[0] + x
+                                tmpy = vector[1] + y
+                                tmpz = vector[2] + z
+                                tmp_tuple = tuple((tmpx, tmpy, tmpz))
+                                value = tuple((the_id, vector))
+                                overlap_dict[tmp_tuple].append(value)
+
```

Comment:
I think you can simplify these three cases into one. Something like below. Also try to avoid `tmp` variables. Be a descriptive as absolutely possible. What does `tmp_tuple` represent physically? 

```python
import itertools as it
for offsets in it.product(range(3), repeat=dimension):
    tmp_tuple = tuple((v + offset for v, offset in zip(vector, offsets))
    overlap_dict[tmp_tuple].append((the_id, vector))
```